### PR TITLE
Scripts/Northrend: Necro Overlord Mezhen

### PR DESCRIPTION
**Pull request by TheVanoq**
**Changes proposed:**
Added spells and behavior to "Necro Lord Mezhen":
- He casts a purple beam onto the structure in front of him
- He casts "Raise the Dead", as well as "Drain Soul" when in combat

**Target branch(es):** 3.3.5/master
- 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
- "Necro Lord Mezhen" had no behavior. He now has an idle cast and the correct combat behavior.

**Tests performed:** (Does it build, tested in-game, etc.)
- Builds and works in-game

**Known issues and TODO list:** (add/remove lines as needed)
- None

### CODE REVIEW

File: sql/updates/world/3.3.5/2017_99_99_99_world.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
...
+	VALUES (24018, 0, 0, 0, 0, 0, 16309, 0, 0, 0, 'Necro Overlord Mezhen', '', '', 0, 71, 71, 2, 21, 0, 1, 1.14286, 1, 0, 0, 2500, 2000, 1, 1, 2, 32768, 2048, 0, 0, 0, 0, 0, 0, 7, 0, 24018, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', 0, 3, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 'npc_necro_overlord_mezhen', 12340);
+
+DELETE FROM conditions WHERE SourceEntry='43559';
+INSERT INTO conditions
+	(SourceTypeOrReferenceId, SourceGroup, SourceEntry, SourceId, ElseGroup, ConditionTypeOrReference, ConditionTarget, ConditionValue1, ConditionValue2, ConditionValue3, NegativeCondition, ErrorType, ErrorTextId, ScriptName, Comment)
+	VALUES (13, 1, 43559, 0, 0, 36, 0, 0, 0, 0, 1, 0, 0, '', '');
```
Comment (Aokromes) [inline sql/updates/world/3.3.5/2017_99_99_99_world.sql]: Don't delete creature template, use update, also missing newline.

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 2]
```diff
@@ -668,4 +760,5 @@ void AddSC_howling_fjord()
     new npc_mindless_abomination();
     new spell_mindless_abomination_explosion_fx_master();
     new npc_riven_widow_cocoon();
- }
+    new npc_necro_overlord_mezhen();
+ }
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: missing newline.

**Comment by TheVanoq** [comment]
Should be all fixed now

**Comment by Killyana** [comment]
It has a quote in aggro:
Necro Overlord Mezhen says: Foolish child, you will not stop what I have set in motion! Now, meet your doom!

**Comment by Treeston** [comment]
Refactor using `EventMap` please - see [boss scripts](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/Naxxramas/boss_faerlina.cpp#L144-L185) for reference.

(NB: `BossAI` has a member `EventMap events` that you'll need to declare yourself.)

PS: Code style - data members below methods.

**Comment by Keader** [comment]
You can use TaskScheduler. Examples in: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Outland/BlackTemple/boss_illidan.cpp#L1371

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 3]
```diff
@@ -15,17 +15,17 @@
...
-SD%Complete: 100
-SDComment: Quest support: 11253, 11241.
-SDCategory: howling_fjord
-EndScriptData */
-
-/* ContentData
-npc_plaguehound_tracker
-npc_apothecary_hanes
-EndContentData */
+ /* ScriptData
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: revert this change

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 4]
```diff
@@ -35,30 +35,30 @@ EndContentData */
 #include "SpellInfo.h"
 #include "SpellScript.h"
 
-/*######
-## npc_apothecary_hanes
-######*/
+ /*######
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: revert this change

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 5]
```diff
@@ -35,30 +35,30 @@ EndContentData */
...
+ ## npc_apothecary_hanes
+ ######*/
 enum Entries
 {
     NPC_APOTHECARY_HANES = 23784,
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: revert this changes

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 6]
```diff
@@ -394,20 +395,20 @@ class npc_razael_and_lyana : public CreatureScript
 
 enum Daegarnn
 {
-    QUEST_DEFEAT_AT_RING            = 11300,
+    QUEST_DEFEAT_AT_RING = 11300,
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: revert this change

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 7]
```diff
@@ -551,61 +552,61 @@ class spell_mindless_abomination_explosion_fx_master : public SpellScriptLoader
...
-    SPELL_SUMMON_FANGGORE_WORG            = 43277,
-    SPELL_SUMMON_GJALERBRON_RUNECASTER    = 43278,
-    SPELL_SUMMON_GJALERBRON_SLEEPWATCHER  = 43279,
-    SPELL_SUMMON_GJALERBRON_WARRIOR       = 43280,
-    SPELL_SUMMON_PUTRID_HORROR            = 43281,
-    SPELL_SUMMON_WINTERSKORN_BERSERKER    = 43282,
-    SPELL_SUMMON_WINTERSKORN_WOODSMAN     = 43283,
-    SPELL_SUMMON_WINTERSKORN_TRIBESMAN    = 43284,
-    SPELL_SUMMON_WINTERSKORN_ORACLE       = 43285,
+    SPELL_SUMMON_BABY_RIVEN_WIDOWS = 43275,
```
Comment (Aokromes) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: revert this changes

**Comment by tkrokli** [comment]
Removing the enum alignment completely is an unwanted codestyle change.
Please restore the alignment for the enum values.

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 8]
```diff
@@ -659,6 +660,111 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+        {
+            events.Reset();
+
+            events.ScheduleEvent(EVENT_CAST_IDLE_BEAM, 333);
+        }
+
+        void EnterCombat(Unit* /*victim*/) override
+        {
+            events.ScheduleEvent(EVENT_CAST_DRAIN_SOUL, randtime(Seconds(0), Seconds(5)));
+            events.ScheduleEvent(EVENT_CAST_RAISE_DEAD, randtime(Seconds(0), Seconds(10)));
```
Comment (sirikfoll) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: you can remove those `randtime`, like this: `events.ScheduleEvent(EVENT_CAST_RAISE_DEAD, Seconds(0), Seconds(10));`

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 9]
```diff
@@ -659,6 +660,111 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+                switch (eid)
+                {
+                    case EVENT_CAST_IDLE_BEAM:
+                        if (me->isMoving())
+                        {
+                            events.Repeat(Milliseconds(3000));
+                            return;
+                        }
+
+                        if (Unit* target = me->FindNearestCreature(24021, 16.0f))
```
Comment (sirikfoll) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: And create a enum to npc 24021

Comment (tkrokli) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: Yes, please avoid "magic numbers" as much as possible (use enum).

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 10]
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+                        me->StopMoving();
+
+                        DoCastVictim(SPELL_DRAIN_SOUL);
+
+                        auto channel_time = Seconds(5);
+                        events.Repeat(randtime(channel_time + Seconds(3), channel_time + Seconds(10)));
+
+                        break;
+                    case EVENT_CAST_RAISE_DEAD:
+                        if (!UpdateVictim())
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: it should be on top of this method.

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 11]
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+                            return;
+
+                        me->StopMoving();
+
+                        DoCast(SPELL_RAISE_DEAD);
+
+                        auto cast_time = Milliseconds(1500);
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: we only use "auto" to iterator

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 12]
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+                        events.Repeat(randtime(cast_time + Seconds(30), cast_time + Seconds(60)));
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: should be: 

``` events.Repeat(Seconds(31)+Milliseconds(500), Seconds(61)+Milliseconds(500));```

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 13]
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+
+    struct npc_necro_overlord_mezhenAI : public ScriptedAI
+    {
+        npc_necro_overlord_mezhenAI(Creature* creature) : ScriptedAI(creature) { }
+
+        void Reset() override
+        {
+            events.Reset();
+
+            events.ScheduleEvent(EVENT_CAST_IDLE_BEAM, 333);
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: events.ScheduleEvent(EVENT_CAST_IDLE_BEAM, Milliseconds(333));

File: src/server/scripts/Northrend/zone_howling_fjord.cpp [hunk 14]
```diff
@@ -659,6 +659,116 @@ class npc_riven_widow_cocoon : public CreatureScript
...
+                    case EVENT_CAST_IDLE_BEAM:
+                        if (me->isMoving())
+                        {
+                            events.Repeat(Seconds(3));
+                            return;
+                        }
+
+                        if (Unit* target = me->FindNearestCreature(BUNNY_BEAM_TARGET, 16.0f))
+                        {
+                            me->CastSpell(target, SPELL_IDLE_PURPLE_BEAM, false);
```
Comment (Keader) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: change it to: DoCast(target, SPELL_IDLE_PURPLE_BEAM);

File: src/server/scripts/Northrend/zone_howling_fjord.cpp (hunk 12 above)
Comment (Warpten) [inline src/server/scripts/Northrend/zone_howling_fjord.cpp]: are you all being serious with your custom std::chrono::fuckwith complaints

**Comment by Treeston** [comment]
@tkrokli please do not make use of "allow edits from maintainers" unless you're in the process of merging a merge-ready PR (to rename SQL files or make very minor changes).

Otherwise, you end up with developers needing to spend valuable time on git merge because somebody felt the need to unilaterally push a whitespace or alignment commit.

**Comment by tkrokli** [comment]
OK. Sorry that I misunderstood that feature. But is that the reason this PR was closed? Should a new PR be started instead?

**Comment by Treeston** [comment]
Please leave PR management to the person making the PR as well as the reviewing devs - wildly firing off commits often makes the process more complicated than it needs to be for the person contributing.

**Comment by Killyana** [comment]
This could be done by SAI:
``` sql
DELETE FROM `creature_text` WHERE `CreatureID` = 24018; -- entire row missing in TDB 335.63
INSERT INTO `creature_text` (`CreatureID`,`GroupID`,`ID`,`Text`,`Type`,`Language`,`Probability`,`Emote`,`Duration`,`Sound`,`BroadcastTextId`,`TextRange`,`comment`) VALUES
(24018, 0, 0, 'Foolish child, you will not stop what I have set in motion! Now, meet your doom!', 12, 0, 100, 0, 0, 0, 23694, 0, 'Necro Overlord Mezhen - SAY_AGGRO');

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=13 AND `SourceEntry` = 43559;
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorType`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
(13, 1, 43559, 0, 0, 36, 0, 0, 0, 0, 1, 0, 0, '', '');

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=22 AND `SourceGroup`=4 AND `SourceEntry` = 24018;
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorType`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
(22, 4, 24018, 0, 0, 29, 1, 24014, 15, 1, 0, 0, 0, '', '');

UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 24018;
DELETE FROM `smart_scripts` WHERE `entryorguid` IN (24018) AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(24018, 0, 0, 0, 1, 0, 100, 1, 1000, 1000, 0, 0, 11, 43151, 2, 0, 0, 0, 0, 11, 24021, 30, 0, 0, 0, 0, 0, "Necro Overlord Mezhen - Out of Combat - Cast 'Necrolord: Purple Beam'(No Repeat)"),
(24018, 0, 1, 0, 4, 0, 100, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, "Necro Overlord Mezhen - Out aggro - say text"),
(24018, 0, 2, 0, 0, 0, 100, 0, 2000, 4000, 3000, 5000, 11, 32862, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, "Necro Overlord Mezhen - IC - cast Drain soul"),
(24018, 0, 3, 4, 0, 0, 100, 0, 8000, 11000, 15000, 20000, 11, 43559, 2, 0, 0, 0, 0, 19, 24014, 15, 1, 0, 0, 0, 0, "Necro Overlord Mezhen - IC - cast RAISE DEAD"),
(24018, 0, 4, 0, 61, 0, 100, 0, 0, 0, 0, 0, 41, 0, 0, 0, 0, 0, 0, 19, 24014, 15, 1, 0, 0, 0, 0, "Necro Overlord Mezhen - IC - Despawn corpse");
